### PR TITLE
Better error detail for setting upgrade status

### PR DESCRIPTION
**Pull request by manadart**
This patch is a follow up to #13058, which had some suggestions post-merge.

It simply mirrors Mongo assertions with real checks inside the transaction source function. This means that we emit real error messages instead of "state changing too quickly".

## QA steps

See unit tests. No functional changes. Test for regression by running a simple upgrade to this patch as per #13058.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: state/upgrade_test.go [hunk 1]
```diff
@@ -490,18 +517,7 @@ func (s *UpgradeSuite) TestSetStatus(c *gc.C) {
...
-	c.Assert(err, jc.ErrorIsNil)
-	assertStatus(state.UpgradeDBComplete)
-
-	err = info.SetStatus(state.UpgradeRunning)
-	c.Assert(err, jc.ErrorIsNil)
-	assertStatus(state.UpgradeRunning)
-	err = info.SetStatus(state.UpgradeRunning)
-	c.Assert(err, jc.ErrorIsNil)
-	assertStatus(state.UpgradeRunning)
+		`upgrade status transition from "pending" to "running" not valid`)
```
Comment (manadart) [inline state/upgrade_test.go:520]: Real effect of this change demonstrated here.

**Comment by manadart** [comment]
[Upgrades tidyup](https://trello.com/c/rc3Zf1uI/3722-upgrades-tidyup)

**Review by SimonRichardson — APPROVED** [review]
Love this change, anything to get rid of at least one "state changing too quickly" error in juju.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**